### PR TITLE
Load cache

### PR DESCRIPTION
**Pull request by markhannum**
This features provides the ability to dump a database's cache-contents to a file, and then use this file to re-populate the cache in a different instance (or the same instance after a bounce).  The goal is to do this automatically upon a clean-exit, and reload it in the new process to reduce the amount of time that an instance runs with a cold cache.

### CODE REVIEW

File: tests/load_cache.test/runit [hunk 1]
```diff
@@ -0,0 +1,272 @@
+#!/usr/bin/env bash
```
Comment (mponomar) [inline tests/load_cache.test/runit]: A lot of code here is useful and not specific to this test (eg bounce_cluster).  Maybe we should move it to a file in tests and source it from here?

**Comment by mponomar** [comment]
Almost forgot - would be nice to have the option to dump/load at start/exit time with a single option.  Load before marking us ready so the cache is warmed up.

**Comment by markhannum** [comment]
Yes- I haven't written that code yet.

**Comment by markhannum** [comment]
The load_cache test failure, the physical replicant failed to start with this message:
[ERROR] net_listen: FAILED TO BIND to port 19000: 98 Address already in use

File: berkdb/mp/mp_sync.c [hunk 2]
```diff
@@ -1057,6 +1172,530 @@ berk_memp_sync_alarm_ms(int x)
...
+	int ret;
+	fileid_page_list_t *pagelist = hash_find(hash, fileid);
+	if (!pagelist) {
+		if ((ret = __os_calloc(dbenv, 1, sizeof(*pagelist), &pagelist)) != 0) {
+			return ret;
+		}
+		memcpy(pagelist->fileid, fileid, DB_FILE_ID_LEN);
+		hash_add(hash, pagelist);
+	}
+	return add_page_to_fileid_page(dbenv, pagelist, pg);
```
Comment (markhannum) [inline berkdb/mp/mp_sync.c]: Maybe change to add_page_to_fileid_list .. ?

File: berkdb/mp/mp_sync.c [hunk 3]
```diff
@@ -1057,6 +1172,530 @@ berk_memp_sync_alarm_ms(int x)
...
+							"%d\n", __func__, ret);
+					return ret;
+				}
+				(*pages)++;
+			}
+			MUTEX_UNLOCK(dbenv, &hp->hash_mutex);
+		}
+	}
+
+    output_fileid_page_hash(dbenv, fileid_pages, s);
```
Comment (markhannum) [inline berkdb/mp/mp_sync.c]: fix tabs

File: berkdb/mp/mp_sync.c [hunk 4]
```diff
@@ -1057,6 +1172,530 @@ berk_memp_sync_alarm_ms(int x)
...
+ *	Write bufferpool fileids and pages to a file
+ *
+ * PUBLIC: int __memp_dump
+ * PUBLIC:	 __P((DB_ENV *, SBUF2 *, u_int32_t *));
+ */
+int
+__memp_dump(dbenv, s, pages)
+	DB_ENV *dbenv;
+	SBUF2 *s;
+	u_int32_t *pages;
```
Comment (markhannum) [inline berkdb/mp/mp_sync.c]: Rename this to "count" or "count_pages", or something more obvious

Comment (markhannum) [inline berkdb/mp/mp_sync.c]: Also maybe make this a u_int64_t

File: berkdb/mp/mp_sync.c [hunk 5]
```diff
@@ -1057,6 +1172,530 @@ berk_memp_sync_alarm_ms(int x)
...
+ *	Load bufferpool fileids and pages to a file
+ *
+ * PUBLIC: int __memp_load
+ * PUBLIC:	 __P((DB_ENV *, SBUF2 *, u_int32_t *, u_int32_t *));
+ */
+int
+__memp_load(dbenv, s, pages, lines)
+	DB_ENV *dbenv;
+	SBUF2 *s;
+	u_int32_t *pages;
```
Comment (markhannum) [inline berkdb/mp/mp_sync.c]: call this count_pages or cout or something .. also make it a u_int64_t

File: berkdb/mp/mp_sync.c [hunk 6]
```diff
@@ -1057,6 +1172,530 @@ berk_memp_sync_alarm_ms(int x)
...
+				break;
+			}
+			pg = hx;
+
+			if (fileid_env == NULL) { 
+				if ((ret = __os_malloc(dbenv, sizeof(*fileid_env), 
+								&fileid_env)) != 0) {
+					logmsg(LOGMSG_ERROR, "%s line %d error alocating memory, %d\n",
+							__func__, __LINE__, ret);
+					return ret;
```
Comment (markhannum) [inline berkdb/mp/mp_sync.c]: Maybe goto done, and wait on previous threads to complete .. ?

File: berkdb/mp/mp_sync.c [hunk 7]
```diff
@@ -1057,6 +1172,530 @@ berk_memp_sync_alarm_ms(int x)
...
+				}
+
+				if ((ret = __os_calloc(dbenv, 1, sizeof(fileid_page_list_t),
+								&fileid_env->pagelist)) != 0) {
+					__os_free(dbenv, fileid_env);
+					logmsg(LOGMSG_ERROR, "%s line %d error alocating memory, %d\n",
+							__func__, __LINE__, ret);
```
Comment (markhannum) [inline berkdb/mp/mp_sync.c]: goto done rather than returning immediately ..

File: berkdb/mp/mp_sync.c [hunk 8]
```diff
@@ -1057,6 +1172,530 @@ berk_memp_sync_alarm_ms(int x)
...
+				fileid_env->lk = &lk;
+				fileid_env->cd = &cd;
+				fileid_env->active_threads = &active_threads;
+				memcpy(fileid_env->pagelist->fileid, fileid, DB_FILE_ID_LEN);
+			}
+
+			if ((ret = add_page_to_fileid_page(dbenv, fileid_env->pagelist, pg)) != 0) {
+				__os_free(dbenv, fileid_env->pagelist);
+				__os_free(dbenv, fileid_env);
+				logmsg(LOGMSG_ERROR, "%s line %d error alocating memory, %d\n",
```
Comment (markhannum) [inline berkdb/mp/mp_sync.c]: goto done

File: berkdb/mp/mp_sync.c [hunk 9]
```diff
@@ -1057,6 +1172,530 @@ berk_memp_sync_alarm_ms(int x)
...
+	}
+	MUTEX_THREAD_UNLOCK(dbenv, dbmp->mutexp);
+
+	if (dbmfp) {
+		for(int pages = 0 ; pages < pagelist->cnt; pages++) {
+			touch_page(dbmfp, pagelist->pages[pages]);
+		}
+	} 
+
+    if (thdpool) {
```
Comment (markhannum) [inline berkdb/mp/mp_sync.c]: fix tabs

File: berkdb/mp/mp_sync.c [hunk 10]
```diff
@@ -1057,6 +1172,530 @@ berk_memp_sync_alarm_ms(int x)
...
+
+static void
+load_fileids_thdpool(fileid_page_env_t *fileid_env)
+{
+	int ret;
+	Pthread_mutex_lock(fileid_env->lk);
+	(*fileid_env->active_threads)++;
+	if ((ret = thdpool_enqueue(loadcache_thdpool, load_fileids,
+					fileid_env, 0, NULL, 0)) != 0) {
+		(*fileid_env->active_threads)--;
```
Comment (markhannum) [inline berkdb/mp/mp_sync.c]: Decrement strictly in the function ..

File: berkdb/mp/mp_sync.c (hunk 9 above)
Comment (markhannum) [inline berkdb/mp/mp_sync.c]: Don't need this if this is the only place that will decrement

File: berkdb/mp/mp_sync.c [hunk 11]
```diff
@@ -1057,6 +1172,530 @@ berk_memp_sync_alarm_ms(int x)
...
+			fileid_env = NULL;
+		}
+		if (!endofline)
+			sbuf2nextline(s);
+	}
+	if (fileid_env) {
+		load_fileids_thdpool(fileid_env);
+		fileid_env = NULL;
+	}
+	Pthread_mutex_lock(&lk);
```
Comment (markhannum) [inline berkdb/mp/mp_sync.c]: add done: label just before this lock

File: berkdb/mp/mp_sync.c [hunk 12]
```diff
@@ -1057,6 +1172,530 @@ berk_memp_sync_alarm_ms(int x)
...
+}
+
+/*
+ * __memp_flush_pagelist --
+ *	Dump bufferpool pagelist to log directory
+ *
+ * PUBLIC: int __memp_flush_pagelist
+ * PUBLIC:	 __P((DB_ENV *));
+ */
+int
```
Comment (markhannum) [inline berkdb/mp/mp_sync.c]: Introduce a tunable that will short-circuit this logic unless a percentage of the bufferpool has change (we can detect misses, and therefore the number of memp_fgets that result in IO, and modify the bufferpool)

File: tests/load_cache.test/runit [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+    typeset file2=$2
+    typeset linecnt
+    typeset faildiff
+    typeset cnt
+    linecnt=$(wc -l ${file1} | awk '{print $1}')
+    [[ -z $linecnt ]] && failexit "Cant read cache file"
+    fd=$(echo "${linecnt} * 0.${failpct}" | bc)
+    faildiff=${fd%%\.*}
+    cat $file1 | while read fileid r ; do for x in $r ; do echo $fileid $x ; done ; done | sort > ${file1}.sort
+    cat $file2 | while read fileid r ; do for x in $r ; do echo $fileid $x ; done ; done | sort > ${file1}.sort
```
Comment (markhannum) [inline tests/load_cache.test/runit]: BUG BUG BUG

File: tests/load_cache.test/runit [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+    [[ $debug == "1" ]] && set -x
+    typeset func="diff_files"
+    write_prompt $func "Running $func"
+    typeset file1=$1
+    typeset file2=$2
+    typeset linecnt
+    typeset faildiff
+    typeset cnt
+    linecnt=$(wc -l ${file1} | awk '{print $1}')
+    [[ -z $linecnt ]] && failexit "Cant read cache file"
```
Comment (markhannum) [inline tests/load_cache.test/runit]: Several problems here .. linecnt should be calculated against the sorted file .. cat file2 should be redirected to file2.sort

**Comment by markhannum** [comment]
cbuild failure seems plugin related:
/bb/bigstorn/systems/gitbuilds/comdb2/opencomdb2/cdb2test/pnj-comdb2btda1/R20190813.14/bb-plugins/explain/explain.c:102:11: error: 'struct sqlclntstate' has no member named 'sql'
     client.sql = sql;

**Pull request merged**